graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -681,21 +684,27 @@ def _compile_output_step(outputs):
         fold_scope_location = output_context['fold']
 
         expression = None
+        existence_check = None
         if fold_scope_location:
-            if optional:
-                raise AssertionError(u'Unreachable state reached, optional in fold: '
-                                     u'{}'.format(output_context))
-
             _, field_name = location.get_location_name()
             expression = expressions.FoldedOutputContextField(
                 fold_scope_location, field_name, graphql_type)
+
+            if optional:
+                existence_check = expressions.ContextFieldExistence(
+                    fold_scope_location.base_location
+                )
         else:
             expression = expressions.OutputContextField(location, graphql_type)
 
             if optional:
                 existence_check = expressions.ContextFieldExistence(location.at_vertex())
-                expression = expressions.TernaryConditional(
-                    existence_check, expression, expressions.NullLiteral)
+
+        # pylint: disable=redefined-variable-type
+        if existence_check:
+            expression = expressions.TernaryConditional(
+                existence_check, expression, expressions.NullLiteral)
+        # pylint: enable=redefined-variable-type
```

Comment:
Overall, let's try to pull as much of this new code out of this one massive function and into helper functions whenever reasonable and possible -- the function is getting pretty unwieldy.

Reply:
Were you referring to `_compile_output_step`? It seems to be doing only one thing?

Reply:
No, the weird folding Github does made me think this was still part of `_compile_vertex_ast` or one of those big functions. No worries, ignore this.